src/Acuminator/Acuminator.Analyzers/StaticAnalysis/CallsToInternalAPI/InternalApiCallsWalker.cs

```diff
@@ -0,0 +1,376 @@
+﻿#nullable enable
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Concurrent;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Runtime.CompilerServices;
+
+using Acuminator.Utilities;
+using Acuminator.Utilities.Common;
+using Acuminator.Utilities.DiagnosticSuppression;
+using Acuminator.Utilities.Roslyn.Constants;
+using Acuminator.Utilities.Roslyn.Semantic;
+using Acuminator.Utilities.Roslyn.Syntax;
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Acuminator.Analyzers.StaticAnalysis.CallsToInternalAPI
+{
+	internal class InternalApiCallsWalker : CSharpSyntaxWalker
+	{
+		private static readonly ConcurrentDictionary<ISymbol, bool> _markedInternalApi = new ConcurrentDictionary<ISymbol, bool>();
```

Comment:
If one solution is closed and another solution is opened, is this collection cleared?
What happens if somebody marked a type as `PXInternalUseOnly`, then the analysis was performed, and then `PXInternalUseOnly` was removed? Won't it lead to a symbol in this collection that has `true` as a value but in fact is no longer marked with `PXInternalUseOnly` in the source code?
What if the symbol was completely removed from the code? May it lead to a memory leak? Have you considered using `ConditionalWeakTable` instead?

Reply:
I was more concerned about parallel access when I wrote this cache and thought about internal API as of something static - that this is a check in ISV mode for external devs who don't have control over internal API and have no reason to use the `PXInternalUseOnly`. Therefore, since the external developer can't change the internal API everything looks not that bad. The cache can be considered as a register for internal/non internal API. 
There is a potential memory leak in the cases you mentioned - when we switch solutions, or make changes to the internal API (imho is a corner case). For the former case I think the `ConditionalWeakTable` is the solution and it is worth to change the code to use it especially since it is thread safe. 

As for the latter case - there is a problem. If we need to consider that the external dev can change the internal API then there is no possibility to cache symbols. We don't explicitly control when the symbol will be deleted from the `ConditionalWeakTable` nor can we somehow know beforehand that internal API was changed and update the cache. It is possible that the developer deleted `PXInternalUseOnly` attribute or the whole API with this attribute but during the analysis it is still there because GC hasn't happened yet.
Maximum we can do in such case is to:
1. Stop caching non internal API which is a performance hit but not a very big one I think
2. Cache only internal API from Acumatica. Check for namespaces for example. It isn't very convenient and we have to either keep a list of namespaces updated or rely on the **PX.** prefix and hope that no one else will use it. There will be small additional overhead of checking namespaces so it is possible that in such case it is best to not cache at all.
Or we can mark `PXInternalUseOnly` attribute itself with `PXInternalUseOnly`  to set in stone that only we control what is an internal API and what isn't.

Reply:
As we discussed previously, our own developer may change `PXInternalUseOnly` marking on some type when working with `PureWithNetTools.sln` or open the same solution from another branch with a different set of `PXInternalUseOnly`-marked types so the cache cannot be global.
I think that, first of all, we need to make sure that there is a performance issue before fixing it in a complex way.

Reply:
I have changed the used cache to `ConditionalWeakTable`. Now even when there are changes to the document the cache is invalidated and old entries are collected. It also works this way when solution changes. The reload of solution also happens when you switch git branches. I have also checked that the diagnostic works with the manual changes of the code like addition of `PXInternalUseOnly` attribute or its removal. We also don't have to worry about multiple instances of VS since each of them has its own memory and therefore cache. Even if the out-of-process feature is turned on there still will be two different code analysis processes for two different Visual Studios
As for the performance issues - it is hard to measure something precisely with diagnostics and it is a separate big amount of work. In my opinion this particular diagnostic should use global caching because:
1. It checks the whole syntax tree, not a particular symbol or node. It works for all types, not just types specific to Acumatica.
2. It works on IdentifierName nodes - the most frequently encountered node in the syntax tree. It also works on other common nodes - GenericName, MemberAccess. By work I mean it makes the most expensive checks here.
3. The checks are not that cheap either. To correctly deduce that the type belongs to the internal API we need to check its base type + its containing type. For each of this two types we also need to check their base type and containing type and etc. As you can see there is an exponential growth of types needed to be checked especially if the type doesn't belong to the internal API (which is the most common case) and we have to check all possible combination of these types. In practice their number is not that big but it still can sum up. For each checked access of a type member we also have to do the same check for its containing type. So the number of checks really grows